File: swift/src/Commit.swift
/// MuonGit - Commit object read/write
import Foundation

/// A parsed git commit object
public struct Commit: Sendable {
    public let oid: OID
    public let treeId: OID
    public let parentIds: [OID]
    public let author: Signature
    public let committer: Signature
    public let message: String
    public let messageEncoding: String?
}

// MARK: - Parsing

/// Parse a commit object from its raw data content
public func parseCommit(oid: OID, data: Data) throws -> Commit {
    guard let text = String(data: data, encoding: .utf8) else {
        throw MuonGitError.invalidObject("commit is not valid UTF-8")
    }

    var treeId: OID?
    var parentIds: [OID] = []
    var author: Signature?
    var committer: Signature?
    var messageEncoding: String?

    // Split into header and message at first blank line
    let parts = text.split(separator: "\n\n", maxSplits: 1)
    let headerSection = String(parts[0])
    let message = parts.count > 1 ? String(parts[1]) : ""

    for line in headerSection.split(separator: "\n", omittingEmptySubsequences: false) {
        let line = String(line)
        if line.hasPrefix("tree ") {
            let hex = String(line.dropFirst(5))
            treeId = OID(hex: hex)
        } else if line.hasPrefix("parent ") {
            let hex = String(line.dropFirst(7))
            parentIds.append(OID(hex: hex))
        } else if line.hasPrefix("author ") {
            author = parseSignature(String(line.dropFirst(7)))
        } else if line.hasPrefix("committer ") {
            committer = parseSignature(String(line.dropFirst(10)))
        } else if line.hasPrefix("encoding ") {
            messageEncoding = String(line.dropFirst(9))
        }
    }

    guard let tree = treeId else {
        throw MuonGitError.invalidObject("commit missing tree")
    }
    guard let auth = author else {
        throw MuonGitError.invalidObject("commit missing author")
    }
    guard let comm = committer else {
        throw MuonGitError.invalidObject("commit missing committer")
    }

    return Commit(
        oid: oid,
        treeId: tree,
        parentIds: parentIds,
        author: auth,
        committer: comm,
        message: message,
        messageEncoding: messageEncoding
    )
}

// MARK: - Serialization

// Pre-computed byte arrays for commit serialization prefixes
private let commitTreePrefix: [UInt8] = [UInt8]("tree ".utf8)
private let commitParentPrefix: [UInt8] = [UInt8]("parent ".utf8)
private let commitAuthorPrefix: [UInt8] = [UInt8]("author ".utf8)
private let commitCommitterPrefix: [UInt8] = [UInt8]("committer ".utf8)
private let commitEncodingPrefix: [UInt8] = [UInt8]("encoding ".utf8)
private let commitSpaceLt: [UInt8] = [UInt8](" <".utf8)
private let commitGtSpace: [UInt8] = [UInt8]("> ".utf8)

/// Serialize a commit to its raw data representation (without the object header)
public func serializeCommit(
    treeId: OID,
    parentIds: [OID],
    author: Signature,
    committer: Signature,
    message: String,
    messageEncoding: String? = nil
) -> Data {
    var buf = [UInt8]()
    buf.reserveCapacity(256 + message.count)

    buf.append(contentsOf: commitTreePrefix)
    treeId.appendHexBytes(to: &buf)
    buf.append(0x0A)

    for pid in parentIds {
        buf.append(contentsOf: commitParentPrefix)
        pid.appendHexBytes(to: &buf)
        buf.append(0x0A)
    }

    buf.append(contentsOf: commitAuthorPrefix)
    appendSignatureBytes(author, to: &buf)
    buf.append(0x0A)

    buf.append(contentsOf: commitCommitterPrefix)
    appendSignatureBytes(committer, to: &buf)
    buf.append(0x0A)

    if let enc = messageEncoding {
        buf.append(contentsOf: commitEncodingPrefix)
        var encStr = enc
        encStr.withUTF8 { buf.append(contentsOf: $0) }
        buf.append(0x0A)
    }

    buf.append(0x0A)
    var msg = message
    msg.withUTF8 { buf.append(contentsOf: $0) }
    return Data(buf)
}

/// Append a signature directly as bytes to a UInt8 buffer
private func appendSignatureBytes(_ sig: Signature, to buf: inout [UInt8]) {
    var name = sig.name
    name.withUTF8 { buf.append(contentsOf: $0) }
    buf.append(contentsOf: commitSpaceLt)
    var email = sig.email
    email.withUTF8 { buf.append(contentsOf: $0) }
    buf.append(contentsOf: commitGtSpace)
    appendInt64(sig.time, to: &buf)
    buf.append(0x20) // space
    buf.append(sig.offset >= 0 ? 0x2B : 0x2D) // '+' or '-'
    let absOffset = abs(sig.offset)
    let hours = absOffset / 60
    let minutes = absOffset % 60
    buf.append(UInt8(hours / 10) + 0x30)
    buf.append(UInt8(hours % 10) + 0x30)
    buf.append(UInt8(minutes / 10) + 0x30)
    buf.append(UInt8(minutes % 10) + 0x30)
}

/// Append decimal representation of an Int64 directly to a UInt8 buffer
private func appendInt64(_ value: Int64, to buf: inout [UInt8]) {
    if value == 0 {
        buf.append(0x30)
        return
    }
    var v = value
    if v < 0 {
        buf.append(0x2D) // '-'
        v = -v
    }
    let start = buf.count
    while v > 0 {
        buf.append(UInt8(v % 10) + 0x30)
        v /= 10
    }
    // Reverse digits in-place
    var lo = start
    var hi = buf.count - 1
    while lo < hi {
        let tmp = buf[lo]
        buf[lo] = buf[hi]
        buf[hi] = tmp
        lo += 1
        hi -= 1
    }
}

// MARK: - Signature helpers

/// Parse "Name <email> timestamp offset" into a Signature
func parseSignature(_ s: String) -> Signature {
    // Format: "Name <email> 1234567890 +0000"
    guard let emailStart = s.firstIndex(of: "<"),
          let emailEnd = s.firstIndex(of: ">") else {
        return Signature(name: s, email: "")
    }

    let name = String(s[s.startIndex..<emailStart]).trimmingCharacters(in: .whitespaces)
    let email = String(s[s.index(after: emailStart)..<emailEnd])

    let remainder = String(s[s.index(after: emailEnd)...]).trimmingCharacters(in: .whitespaces)
    let parts = remainder.split(separator: " ")

    var time: Int64 = 0
    var offset: Int32 = 0
    if parts.count >= 1, let t = Int64(parts[0]) {
        time = t
    }
    if parts.count >= 2 {
        offset = parseTimezoneOffset(String(parts[1]))
    }

    return Signature(name: name, email: email, time: time, offset: offset)
}

/// Format a Signature into "Name <email> timestamp offset"
func formatSignature(_ sig: Signature) -> String {
    let sign = sig.offset >= 0 ? "+" : "-"
    let absOffset = abs(sig.offset)
    let hours = absOffset / 60
    let minutes = absOffset % 60
    return "\(sig.name) <\(sig.email)> \(sig.time) \(sign)\(String(format: "%02d%02d", hours, minutes))"
}

/// Parse "+0530" or "-0800" into minutes offset
func parseTimezoneOffset(_ s: String) -> Int32 {
    guard s.count >= 5 else { return 0 }
    let sign: Int32 = s.hasPrefix("-") ? -1 : 1
    let digits = String(s.dropFirst())
    guard digits.count == 4,
          let hours = Int32(digits.prefix(2)),
          let minutes = Int32(digits.suffix(2)) else { return 0 }
    return sign * (hours * 60 + minutes)
}

File: rust/src/commit.rs
//! Commit object read/write

use crate::error::MuonGitError;
use crate::oid::OID;
use crate::types::{ObjectType, Signature};

/// A parsed git commit object
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Commit {
    pub oid: OID,
    pub tree_id: OID,
    pub parent_ids: Vec<OID>,
    pub author: Signature,
    pub committer: Signature,
    pub message: String,
    pub message_encoding: Option<String>,
}

/// Parse a commit object from its raw data content
pub fn parse_commit(oid: OID, data: &[u8]) -> Result<Commit, MuonGitError> {
    let text = std::str::from_utf8(data)
        .map_err(|_| MuonGitError::InvalidObject("commit is not valid UTF-8".into()))?;

    let mut tree_id: Option<OID> = None;
    let mut parent_ids: Vec<OID> = Vec::new();
    let mut author: Option<Signature> = None;
    let mut committer: Option<Signature> = None;
    let mut message_encoding: Option<String> = None;

    // Split at first blank line
    let (header_section, message) = match text.find("\n\n") {
        Some(idx) => (&text[..idx], &text[idx + 2..]),
        None => (text, ""),
    };

    for line in header_section.split('\n') {
        if let Some(hex) = line.strip_prefix("tree ") {
            tree_id = Some(OID::from_hex(hex)?);
        } else if let Some(hex) = line.strip_prefix("parent ") {
            parent_ids.push(OID::from_hex(hex)?);
        } else if let Some(sig_str) = line.strip_prefix("author ") {
            author = Some(parse_signature(sig_str));
        } else if let Some(sig_str) = line.strip_prefix("committer ") {
            committer = Some(parse_signature(sig_str));
        } else if let Some(enc) = line.strip_prefix("encoding ") {
            message_encoding = Some(enc.to_string());
        }
    }

    Ok(Commit {
        oid,
        tree_id: tree_id
            .ok_or_else(|| MuonGitError::InvalidObject("commit missing tree".into()))?,
        parent_ids,
        author: author
            .ok_or_else(|| MuonGitError::InvalidObject("commit missing author".into()))?,
        committer: committer
            .ok_or_else(|| MuonGitError::InvalidObject("commit missing committer".into()))?,
        message: message.to_string(),
        message_encoding,
    })
}

/// Serialize a commit to its raw data representation (without the object header)
pub fn serialize_commit(
    tree_id: &OID,
    parent_ids: &[OID],
    author: &Signature,
    committer: &Signature,
    message: &str,
    message_encoding: Option<&str>,
) -> Vec<u8> {
    let mut buf = String::new();
    buf.push_str(&format!("tree {}\n", tree_id.hex()));
    for pid in parent_ids {
        buf.push_str(&format!("parent {}\n", pid.hex()));
    }
    buf.push_str(&format!("author {}\n", format_signature(author)));
    buf.push_str(&format!("committer {}\n", format_signature(committer)));
    if let Some(enc) = message_encoding {
        buf.push_str(&format!("encoding {}\n", enc));
    }
    buf.push('\n');
    buf.push_str(message);
    buf.into_bytes()
}

/// Parse "Name <email> timestamp offset" into a Signature
fn parse_signature(s: &str) -> Signature {
    let email_start = match s.find('<') {
        Some(i) => i,
        None => return Signature { name: s.to_string(), email: String::new(), time: 0, offset: 0 },
    };
    let email_end = match s.find('>') {
        Some(i) => i,
        None => return Signature { name: s.to_string(), email: String::new(), time: 0, offset: 0 },
    };

    let name = s[..email_start].trim().to_string();
    let email = s[email_start + 1..email_end].to_string();
    let remainder = s[email_end + 1..].trim();
    let parts: Vec<&str> = remainder.split_whitespace().collect();

    let time = parts.first().and_then(|s| s.parse::<i64>().ok()).unwrap_or(0);
    let offset = parts.get(1).map(|s| parse_timezone_offset(s)).unwrap_or(0);

    Signature { name, email, time, offset }
}

/// Format a Signature into "Name <email> timestamp offset"
pub fn format_signature(sig: &Signature) -> String {
    let sign = if sig.offset >= 0 { "+" } else { "-" };
    let abs = sig.offset.unsigned_abs();
    let hours = abs / 60;
    let minutes = abs % 60;
    format!("{} <{}> {} {}{:02}{:02}", sig.name, sig.email, sig.time, sign, hours, minutes)
}

/// Parse "+0530" or "-0800" into minutes offset
fn parse_timezone_offset(s: &str) -> i32 {
    if s.len() < 5 {
        return 0;
    }
    let sign: i32 = if s.starts_with('-') { -1 } else { 1 };
    let digits = &s[1..];
    if digits.len() != 4 {
        return 0;
    }
    let hours: i32 = digits[..2].parse().unwrap_or(0);
    let minutes: i32 = digits[2..].parse().unwrap_or(0);
    sign * (hours * 60 + minutes)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_signature() {
        let sig = parse_signature("Test User <test@example.com> 1234567890 +0530");
        assert_eq!(sig.name, "Test User");
        assert_eq!(sig.email, "test@example.com");
        assert_eq!(sig.time, 1234567890);
        assert_eq!(sig.offset, 330); // 5*60+30
    }

    #[test]
    fn test_format_signature() {
        let sig = Signature {
            name: "Test".into(),
            email: "test@test.com".into(),
            time: 1000,
            offset: -480,
        };
        assert_eq!(format_signature(&sig), "Test <test@test.com> 1000 -0800");
    }

    #[test]
    fn test_parse_and_serialize_commit() {
        let tree = OID::from_hex("4b825dc642cb6eb9a060e54bf899d69f7cb46237").unwrap();
        let author = Signature {
            name: "Author".into(),
            email: "author@example.com".into(),
            time: 1234567890,
            offset: 0,
        };
        let committer = Signature {
            name: "Committer".into(),
            email: "committer@example.com".into(),
            time: 1234567890,
            offset: 0,
        };

        let data = serialize_commit(&tree, &[], &author, &committer, "Initial commit\n", None);
        let oid = OID::hash_object(ObjectType::Commit, &data);
        let commit = parse_commit(oid.clone(), &data).unwrap();

        assert_eq!(commit.tree_id, tree);
        assert!(commit.parent_ids.is_empty());
        assert_eq!(commit.author.name, "Author");
        assert_eq!(commit.committer.email, "committer@example.com");
        assert_eq!(commit.message, "Initial commit\n");
        assert!(commit.message_encoding.is_none());
    }

    #[test]
    fn test_parse_commit_with_parents() {
        let tree = OID::from_hex("4b825dc642cb6eb9a060e54bf899d69f7cb46237").unwrap();
        let parent1 = OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap();
        let parent2 = OID::from_hex("bbf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap();
        let sig = Signature {
            name: "Test".into(),
            email: "test@test.com".into(),
            time: 0,
            offset: 0,
        };

        let data = serialize_commit(&tree, &[parent1.clone(), parent2.clone()], &sig, &sig, "merge\n", None);
        let oid = OID::hash_object(ObjectType::Commit, &data);
        let commit = parse_commit(oid, &data).unwrap();

        assert_eq!(commit.parent_ids.len(), 2);
        assert_eq!(commit.parent_ids[0], parent1);
        assert_eq!(commit.parent_ids[1], parent2);
    }

    #[test]
    fn test_parse_commit_missing_tree() {
        let data = b"author Test <t@t.com> 0 +0000\ncommitter Test <t@t.com> 0 +0000\n\nmsg\n";
        let oid = OID::zero();
        assert!(parse_commit(oid, data).is_err());
    }

    #[test]
    fn test_commit_roundtrip_with_encoding() {
        let tree = OID::from_hex("4b825dc642cb6eb9a060e54bf899d69f7cb46237").unwrap();
        let sig = Signature {
            name: "Test".into(),
            email: "test@test.com".into(),
            time: 100,
            offset: 0,
        };
        let data = serialize_commit(&tree, &[], &sig, &sig, "msg\n", Some("UTF-8"));
        let oid = OID::hash_object(ObjectType::Commit, &data);
        let commit = parse_commit(oid, &data).unwrap();
        assert_eq!(commit.message_encoding.as_deref(), Some("UTF-8"));
    }
}
